### DISCUSSION TITLE
Alternative idea for primary constructor bodies

### DISCUSSION THREAD

**Discussion by boriseetgerink** [dotnet/csharplang#6518] [Language Ideas]
Instead of a statement block as proposed:
```
public class C(int i, string s) : B(s)
{
    {
        if (i < 0) throw new ArgumentOutOfRangeException(nameof(i));
    }
	int[] a = new int[i];
    public int S => s;
}
```

Why not leverage the init keyword? The name of the private property would have to match the name of the primary constructor parameter:

```
public class C(string s)
{
    private readonly int _length;

    private string s {
        init {
            if (string.IsNullOrEmpty(value)) {
                throw new ArgumentNullException(nameof(s));
            }

            _length = value.Length;
        }
    }
}
```

Validation of the constructor parameter is possible, even when not capturing `s` in this case.

A drawback I see is that validating a combination of parameters is not possible, but you could use a normal constructor if you want more advanced validation.

**Comment by symbiogenesis** [comment]
Kotlin uses "initializer blocks" that use the init keyword, but it looks like this:

```kotlin
class InitOrderDemo(name: String) {
    val firstProperty = "First property: $name".also(::println)
    
    init {
        println("First initializer block that prints $name")
    }
    
    val secondProperty = "Second property: ${name.length}".also(::println)
    
    init {
        println("Second initializer block that prints ${name.length}")
    }
}
```

https://kotlinlang.org/docs/classes.html#constructors

**Comment by 333fred** [comment]
Not sure where you're seeing that as the proposed syntax. There is currently no proposal for primary constructor bodies, and the feedback the last time we tried to ship that syntax (C# 6) was so bad it was pulled just before release.

**Reply by HaloFour** [reply]
IIRC the C# 6 proposal co-opted the body of the type as the primary constructor body, allowing for arbitrary statements to be mixed in amongst the member declarations and causing all kinds of weirdness.  I think something a lot simpler, like an `init` block, would be a lot more palatable.

It's probably come up because it was mentioned here: https://github.com/dotnet/csharplang/discussions/7377#discussioncomment-6541283

**Reply by symbiogenesis** [reply]
The example code in the original post is the same code that is shown here, under "possible extensions"

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/primary-constructors#possible-extensions
